client/client_pool/include/client/client_pool/client_pool_timer.hpp

```diff
@@ -15,49 +15,45 @@
 
 #include <chrono>
 #include <future>
-#include <stdexcept>
 
-#include <boost/asio/deadline_timer.hpp>
-#include <boost/asio/io_service.hpp>
-#include <boost/date_time/posix_time/posix_time_duration.hpp>
-#include <boost/system/error_code.hpp>
+#include <asio/steady_timer.hpp>
+#include <asio/io_context.hpp>
 
 namespace concord_client_pool {
 template <typename ClientT>
 class Timer {
  public:
   using Clock = std::chrono::high_resolution_clock;
   Timer(std::chrono::milliseconds timeout, std::function<void(ClientT&&)> on_timeout)
-      : timeout_{timeout}, on_timeout_{on_timeout}, timer_(io_service_) {
+      : timeout_{timeout}, on_timeout_{on_timeout}, timer_(io_context_) {
     if (timeout_.count() > 0) {
-      timer_thread_future_ = std::async(std::launch::async, &Timer::WorkerThread, this);
+      timer_thread_future_ = std::async(std::launch::async, &Timer::work, this);
     }
   }
 
-  ~Timer() { stop(); }
+  ~Timer() { stopTimerThread(); }
 
-  void stop() {
+  void stopTimerThread() {
     if (timeout_.count() == 0) {
       return;
     }
-    stop_ = true;
-    io_service_.stop();
+    io_context_.stop();
```

Comment:
Who will re-run io_context_ after client_pool will call stopTimerThread function?

Reply:
No one. The point of `stopTimerThread()` is to stop the timer thread and hence the io_context. If you want to restart the thread (and start the io_context) then you need to create a new `Timer` object. That's how the implementation was done and I didn't change it. I agree that this could be improved but I didn't want to change the implementation for no reason so I would leave this to a future task if there is a need.

Reply:
That's not what I see in the code. batch_timer_ object is created only in ConcordClientPool ctor.

Reply:
`start()` to start the timeout and restart it for a new client. `cancel()` to cancel the current timeout. `stopTimerThread()` to stop the timer thread.